[AMDGPU] Fix mode register pass for constrained FP operations

abhigargrepo opened the pull request:
This PR will fix the si-mode-register pass which is inserting an extra setreg instruction in case of constrained FP operations. This pass will be ignored for strictfp functions.

abhigargrepo committed:
[AMDGPU] Fix mode register pass for constrained FP operations

This PR will fix the si-mode-register pass which is inserting an extra setreg instruction
in case of constrained FP operations. This pass will be ignored for strictfp functions.

arsenm commented on llvm/test/CodeGen/AMDGPU/mode-register-fpconstrain.ll:
```diff
@@ -9,8 +9,7 @@ define double @ignoreStrictfp(double noundef %a, double noundef %b) #0 {
 ; GCN-NEXT:    s_setreg_imm32_b32 hwreg(HW_REG_MODE, 2, 2), 1
-; GCN-NEXT:    s_nop 1
-; GCN-NEXT:    s_setreg_imm32_b32 hwreg(HW_REG_MODE, 2, 1), 0
+; GCN-NOT:     s_setreg_imm32_b32 hwreg(HW_REG_MODE, 2, 1), 0
```
Don't manually add the -NOT here. The generated -NEXT checks are sufficient and much more reliable

arsenm commented on llvm/lib/Target/AMDGPU/SIModeRegister.cpp:
```diff
@@ -430,6 +430,9 @@ void SIModeRegister::processBlockPhase3(MachineBasicBlock &MBB,
 
 bool SIModeRegister::runOnMachineFunction(MachineFunction &MF) {
+  const Function &F = MF.getFunction();
+  if (F.hasFnAttribute(llvm::Attribute::StrictFP))
```
Add a comment explaining why

abhigargrepo replied:
Will update

arsenm replied:
This comment doesn't really explain why though. Should say something about how strictfp functions have user managed FP modes

abhigargrepo replied:
Updated, kindly review

abhigargrepo committed:
Merge branch 'main' into fixBug-siMode

abhigargrepo committed:
[AMDGPU] Fix mode register pass for constrained FP operations

This PR will fix the si-mode-register pass which is inserting an extra setreg instruction
in case of constrained FP operations. This pass will be ignored for strictfp functions.

abhigargrepo committed:
Merge branch 'main' into fixBug-siMode

abhigargrepo committed:
[AMDGPU] Fix mode register pass for constrained FP operations

This PR will fix the si-mode-register pass which is inserting an extra setreg instruction in case of constrained FP operations.
This pass will be ignored for strictfp functions.
